File: apps/daemon/src/coagentia_daemon/adapters/frames.py
"""stream-json 帧 → 上报回调映射（契约 E §7/§8；锚定 CLI 2.1.205 真机帧）。

纯逻辑，无子进程依赖——可用桩帧全量单测（防腐层 / 相位聚合 / usage 提取 / 诊断映射）。

**帧防腐层（铁律 4）**：未知 top-level `type` / 未知 `system.subtype` / 契约外帧
（rate_limit_event、system/status|api_retry|notification）一律忽略并计数；每种未知类型**首现**
写一条低频 `agent.unknown_frame` 诊断（后续同类型静默累加 `unknown_counts`）——CLI 升级不崩。

**相位聚合（§7.2/§7.3）**：仅在相位切换时回调一帧 activity；同相位 delta 帧不产生上报。
**usage 提取（§7.4）**：唯一提取点 = result 帧；id=适配器 ULID（exactly-once 去重根基），
source_session=init 帧的 session_id（UUID）；忽略 message_delta 中间 usage 与 total_cost_usd。
"""

from __future__ import annotations

from collections.abc import Callable
from typing import Any

from coagentia_contracts.daemon import DiagnosticEventIn, TokenUsageEventIn
from coagentia_contracts.enums import AgentStatus

from coagentia_daemon.adapter import AdapterSink
from coagentia_daemon.util import new_ulid, now_iso

# 相位文案（ACTIVITY_PHRASES 值域，契约 E §7.2 / C §6.2）
_P_THINKING = "Thinking…"
_P_REPLYING = "Replying…"
_P_COMMAND = "Running command…"
_P_WRITING = "Writing file…"
_P_READING = "Reading files…"
_P_BROWSING = "Browsing…"
_P_SUBAGENT = "Subagent started"
_P_USING = "Using {tool}…"  # ACTIVITY_PHRASES 模板

# 工具名 → 相位分类（其余工具落 "Using {tool}…"）
_TOOLS_COMMAND = frozenset({"Bash", "BashOutput", "KillShell", "KillBash"})
_TOOLS_WRITING = frozenset({"Edit", "Write", "MultiEdit", "NotebookEdit"})
_TOOLS_READING = frozenset({"Read", "Glob", "Grep", "LS", "NotebookRead"})
_TOOLS_BROWSING = frozenset({"WebFetch", "WebSearch"})
_TOOLS_SUBAGENT = frozenset({"Task"})

_TURN_OUTPUT_PREVIEW_MAX = 500  # §8 agent.turn_output 正文预览上限
_COMMAND_PREVIEW_MAX = 500


def _short_tool(name: str) -> str:
    """MCP 工具名 mcp__coagentia__send_message → send_message（活动文案可读性）。"""
    if name.startswith("mcp__"):
        return name.rsplit("__", 1)[-1] or name
    return name


def phase_for_tool(name: str) -> str:
    """工具名 → activity 相位（契约 E §7.2）。"""
    if name in _TOOLS_COMMAND:
        return _P_COMMAND
    if name in _TOOLS_WRITING:
        return _P_WRITING
    if name in _TOOLS_READING:
        return _P_READING
    if name in _TOOLS_BROWSING:
        return _P_BROWSING
    if name in _TOOLS_SUBAGENT:
        return _P_SUBAGENT
    return _P_USING.format(tool=_short_tool(name))


class FrameRouter:
    """单 Agent 的 stream-json 帧解析器 → AdapterSink 四回调。

    每进程一实例；turn 上下文（channel_id/thread_root_id）由管理器在 feed 前置入，
    result 帧据此为 usage 事件打归属提示（server 富化为 task_id）。
    """

    def __init__(
        self,
        agent_member_id: str,
        sink: AdapterSink,
        *,
        ulid: Callable[[], str] = new_ulid,
        now: Callable[[], str] = now_iso,
        on_session: Callable[[str], None] | None = None,
    ) -> None:
        self.agent_member_id = agent_member_id
        self._sink = sink
        self._ulid = ulid
        self._now = now
        self._on_session = on_session
        # 会话簿记
        self.session_id: str | None = None
        self.model: str | None = None
        # 相位聚合
        self._phase: str | None = None
        # turn 进行中标记（feed 置位；result 复位）——防止 init 帧在 turn 中途误报 idle。
        # 实测本 CLI stream-json 输入模式：init 帧**首个 stdin 输入后**才到（E §11.3），
        # 故 init→idle 需避开正在进行的 turn。
        self._turn_active = False
        # 会话已确认（见过 init 或 result）——resume 是否真正生效的判据（管理器降级用）。
        self.confirmed = False
        # 防腐层计数（lifetime）
        self.unknown_counts: dict[str, int] = {}
        # tool_use_id → (name, input) 关联（assistant/stream_event 登记，user tool_result 回查）
        self._tool_uses: dict[str, tuple[str, dict[str, Any]]] = {}
        # turn 归属提示
        self.channel_id: str | None = None
        self.thread_root_id: str | None = None
        # 最近一次上报状态（管理器读，避免重复 emit）
        self.last_status: AgentStatus | None = None

    def set_turn_context(self, channel_id: str | None, thread_root_id: str | None) -> None:
        self.channel_id = channel_id
        self.thread_root_id = thread_root_id

    def reset_phase(self) -> None:
        self._phase = None

    def begin_turn(self) -> None:
        """管理器喂入一个 turn 前调用：标记 turn 进行中（init→idle 抑制窗口）。"""
        self._turn_active = True

    def reset_run(self) -> None:
        """一次 spawn 的运行态复位（confirmed/turn 标记；session_id 保留）。"""
        self.confirmed = False
        self._turn_active = False
        self._phase = None

    # ------------------------------------------------------------ 分发

    async def process(self, frame: dict[str, Any]) -> None:
        """路由一帧（防腐：任何未知/畸形帧不得抛出到调用方）。"""
        ftype = frame.get("type")
        if ftype == "system":
            await self._on_system(frame)
        elif ftype == "stream_event":
            await self._on_stream_event(frame)
        elif ftype == "assistant":
            await self._on_assistant(frame)
        elif ftype == "user":
            await self._on_user(frame)
        elif ftype == "result":
            await self._on_result(frame)
        else:
            self._count_unknown(str(ftype))

    # ------------------------------------------------------------ system（init → idle）

    async def _on_system(self, frame: dict[str, Any]) -> None:
        subtype = frame.get("subtype")
        if subtype == "init":
            sid = frame.get("session_id")
            if sid and sid != self.session_id:
                self.session_id = sid
                if self._on_session is not None:
                    self._on_session(sid)  # 会话簿记（daemon/state/<id>.json，§4）
            self.model = frame.get("model") or self.model
            self.confirmed = True  # resume 生效 / 会话就绪的权威凭据
            if not self._turn_active:
                # turn 未在跑（罕见：某些模式 init 先于输入）→ 就绪 idle。
                await self._status(AgentStatus.IDLE)
        else:
            # system/status | api_retry | notification（契约外噪声）→ 防腐层
            self._count_unknown(f"system/{subtype}")

    # ------------------------------------------------------------ stream_event（相位聚合）

    async def _on_stream_event(self, frame: dict[str, Any]) -> None:
        event = frame.get("event") or {}
        etype = event.get("type")
        if etype == "content_block_start":
            block = event.get("content_block") or {}
            btype = block.get("type")
            if btype == "thinking":
                await self._switch_phase(_P_THINKING)
            elif btype == "text":
                await self._switch_phase(_P_REPLYING)
            elif btype == "tool_use":
                name = block.get("name") or ""
                bid = block.get("id")
                if bid:
                    self._tool_uses[bid] = (name, block.get("input") or {})
                await self._switch_phase(phase_for_tool(name))
        # message_start / content_block_delta / content_block_stop / message_delta /
        # message_stop：同相位或无相位 → 不上报（§7.2）

    async def _switch_phase(self, phase: str) -> None:
        if phase == self._phase:
            return
        self._phase = phase
        await self._sink.on_activity(self.agent_member_id, phase)

    # ------------------------------------------------------------ assistant（turn_output 诊断）

    async def _on_assistant(self, frame: dict[str, Any]) -> None:
        msg = frame.get("message") or {}
        content = msg.get("content") or []
        text_parts: list[str] = []
        tool_calls = 0
        for block in content:
            btype = block.get("type")
            if btype == "text":
                text_parts.append(block.get("text") or "")
            elif btype == "tool_use":
                tool_calls += 1
                bid = block.get("id")
                if bid:
                    self._tool_uses[bid] = (block.get("name") or "", block.get("input") or {})
        preview = "".join(text_parts)[:_TURN_OUTPUT_PREVIEW_MAX]
        self._diag(
            "agent.turn_output",
            {
                "preview": preview,  # 正文**不外发**，仅截断留痕（铁律 5）
                "tool_calls": tool_calls,
                "stop_reason": msg.get("stop_reason"),
            },
        )

    # ------------------------------------------------------------ user（tool_result → 诊断）

    async def _on_user(self, frame: dict[str, Any]) -> None:
        msg = frame.get("message") or {}
        content = msg.get("content") or []
        if not isinstance(content, list):
            return
        for block in content:
            if not isinstance(block, dict) or block.get("type") != "tool_result":
                continue
            tid = str(block.get("tool_use_id") or "")  # 缺失/非法 → "" 查默认（帧防腐）
            name, tool_input = self._tool_uses.get(tid, ("", {}))
            is_error = bool(block.get("is_error"))
            if name in _TOOLS_COMMAND:
                cmd = str(tool_input.get("command", ""))[:_COMMAND_PREVIEW_MAX]
                self._diag("agent.command", {"command": cmd, "is_error": is_error})
            elif name in _TOOLS_WRITING:
                kind = "create" if name == "Write" else "edit"
                self._diag(
                    "agent.file_edit",
                    {"path": tool_input.get("file_path"), "kind": kind, "is_error": is_error},
                )
            else:
                self._diag("agent.tool_call", {"tool": name or "unknown", "ok": not is_error})

    # ------------------------------------------------------------ result（usage + idle/error）

    async def _on_result(self, frame: dict[str, Any]) -> None:
        usage = frame.get("usage") or {}
        event = TokenUsageEventIn(
            id=self._ulid(),  # 每 result 帧一个 ULID（exactly-once 去重根基，§7.4）
            agent_member_id=self.agent_member_id,
            channel_id=self.channel_id,
            thread_root_id=self.thread_root_id,
            input_tokens=int(usage.get("input_tokens", 0) or 0),
            output_tokens=int(usage.get("output_tokens", 0) or 0),
            cache_read_tokens=int(usage.get("cache_read_input_tokens", 0) or 0),
            cache_write_tokens=int(usage.get("cache_creation_input_tokens", 0) or 0),
            source_session=self.session_id,
            reported_at=self._now(),
        )
        self._sink.on_usage(event)  # total_cost_usd / modelUsage.costUSD 忽略（永不折算货币）

        self.confirmed = True
        self._turn_active = False
        self.reset_phase()
        self._tool_uses.clear()
        subtype = frame.get("subtype")
        is_error = bool(frame.get("is_error")) or bool(frame.get("api_error_status"))
        if subtype == "success" and not is_error:
            await self._status(AgentStatus.IDLE)
        else:
            detail = str(subtype or frame.get("api_error_status") or "result_error")
            await self._status(AgentStatus.ERROR, error_detail=detail)

    # ------------------------------------------------------------ 防腐层

    def _count_unknown(self, key: str) -> None:
        seen = self.unknown_counts.get(key, 0)
        self.unknown_counts[key] = seen + 1
        if seen == 0:  # 每种未知类型首现写一条低频诊断，后续静默累加
            self._diag("agent.unknown_frame", {"type": key, "count": 1})

    # ------------------------------------------------------------ 回调底座

    async def _status(self, status: AgentStatus, error_detail: str | None = None) -> None:
        self.last_status = status
        await self._sink.on_status_changed(self.agent_member_id, status, error_detail)

    def _diag(self, dtype: str, payload: dict[str, Any]) -> None:
        self._sink.on_diagnostic(
            DiagnosticEventIn(
                agent_member_id=self.agent_member_id,
                type=dtype,
                channel_id=self.channel_id,
                payload=payload,
                at=self._now(),
            )
        )

File: apps/daemon-ts/src/adapters/frames.ts
/**
 * stream-json 帧 → 上报回调映射（契约 E §7/§8；锚定 CLI 2.1.205 真机帧）。
 *
 * 纯逻辑，无子进程依赖——可用桩帧全量单测（防腐层 / 相位聚合 / usage 提取 / 诊断映射）。
 *
 * **帧防腐层（铁律 4）**：未知 top-level `type` / 未知 `system.subtype` / 契约外帧
 * （rate_limit_event、system/status|api_retry|notification）一律忽略并计数；每种未知类型**首现**
 * 写一条低频 `agent.unknown_frame` 诊断（后续同类型静默累加 `unknownCounts`）——CLI 升级不崩。
 *
 * **相位聚合（§7.2/§7.3）**：仅在相位切换时回调一帧 activity；同相位 delta 帧不产生上报。
 * **usage 提取（§7.4）**：唯一提取点 = result 帧；id=适配器 ULID（exactly-once 去重根基），
 * source_session=init 帧的 session_id（UUID）；忽略 message_delta 中间 usage 与 total_cost_usd。
 *
 * 对等基准 = apps/daemon adapters/frames.py。
 */

import type { AgentStatus, TokenUsageEventIn } from '@coagentia/contracts-ts';

import type { AdapterSink } from '../adapter.ts';
import type { JsonObject } from '../protocol.ts';
import { newUlid, nowIso } from '../util.ts';

// 相位文案（ACTIVITY_PHRASES 值域，契约 E §7.2 / C §6.2）。
// py 侧为模块私有 `_P_*`，但 codex.py 跨模块消费 → TS 侧导出（W3 codex.ts 消费面）。
export const P_THINKING = 'Thinking…';
export const P_REPLYING = 'Replying…';
export const P_COMMAND = 'Running command…';
export const P_WRITING = 'Writing file…';
export const P_READING = 'Reading files…';
export const P_BROWSING = 'Browsing…';
export const P_SUBAGENT = 'Subagent started';
export const P_USING = 'Using {tool}…'; // ACTIVITY_PHRASES 模板

// 工具名 → 相位分类（其余工具落 "Using {tool}…"）
const TOOLS_COMMAND = new Set(['Bash', 'BashOutput', 'KillShell', 'KillBash']);
const TOOLS_WRITING = new Set(['Edit', 'Write', 'MultiEdit', 'NotebookEdit']);
const TOOLS_READING = new Set(['Read', 'Glob', 'Grep', 'LS', 'NotebookRead']);
const TOOLS_BROWSING = new Set(['WebFetch', 'WebSearch']);
const TOOLS_SUBAGENT = new Set(['Task']);

const TURN_OUTPUT_PREVIEW_MAX = 500; // §8 agent.turn_output 正文预览上限
const COMMAND_PREVIEW_MAX = 500;

/** MCP 工具名 mcp__coagentia__send_message → send_message（活动文案可读性）。 */
export function shortTool(name: string): string {
  if (name.startsWith('mcp__')) {
    const idx = name.lastIndexOf('__');
    return name.slice(idx + 2) || name;
  }
  return name;
}

/** 工具名 → activity 相位（契约 E §7.2）。 */
export function phaseForTool(name: string): string {
  if (TOOLS_COMMAND.has(name)) return P_COMMAND;
  if (TOOLS_WRITING.has(name)) return P_WRITING;
  if (TOOLS_READING.has(name)) return P_READING;
  if (TOOLS_BROWSING.has(name)) return P_BROWSING;
  if (TOOLS_SUBAGENT.has(name)) return P_SUBAGENT;
  return P_USING.replace('{tool}', shortTool(name));
}

function isDict(v: unknown): v is JsonObject {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}

/** py `int(x or 0)` 对等：None/缺失/假值 → 0；数值截断取整（非法值 TS 侧落 0 不抛，防腐同族）。 */
function asInt(v: unknown): number {
  const n = Number(v ?? 0);
  return Number.isFinite(n) ? Math.trunc(n) : 0;
}

export interface FrameRouterOptions {
  ulid?: () => string;
  now?: () => string;
  onSession?: ((sessionId: string) => void) | null;
}

/**
 * 单 Agent 的 stream-json 帧解析器 → AdapterSink 四回调。
 *
 * 每进程一实例；turn 上下文（channelId/threadRootId）由管理器在 feed 前置入，
 * result 帧据此为 usage 事件打归属提示（server 富化为 task_id）。
 */
export class FrameRouter {
  readonly agentMemberId: string;
  private readonly sink: AdapterSink;
  private readonly ulid: () => string;
  private readonly now: () => string;
  private readonly onSession: ((sessionId: string) => void) | null;

  // 会话簿记
  sessionId: string | null = null;
  model: string | null = null;
  // 相位聚合
  private phase: string | null = null;
  // turn 进行中标记（beginTurn 置位；result 复位）——防止 init 帧在 turn 中途误报 idle。
  // 实测本 CLI stream-json 输入模式：init 帧**首个 stdin 输入后**才到（E §11.3），
  // 故 init→idle 需避开正在进行的 turn。
  private turnActive = false;
  // 会话已确认（见过 init 或 result）——resume 是否真正生效的判据（管理器降级用）。
  confirmed = false;
  // 防腐层计数（lifetime）
  unknownCounts: Record<string, number> = {};
  // tool_use_id → [name, input] 关联（assistant/stream_event 登记，user tool_result 回查）
  private toolUses = new Map<string, [string, JsonObject]>();
  // turn 归属提示
  channelId: string | null = null;
  threadRootId: string | null = null;
  // 最近一次上报状态（管理器读，避免重复 emit）
  lastStatus: AgentStatus | null = null;

  constructor(agentMemberId: string, sink: AdapterSink, opts: FrameRouterOptions = {}) {
    this.agentMemberId = agentMemberId;
    this.sink = sink;
    this.ulid = opts.ulid ?? newUlid;
    this.now = opts.now ?? nowIso;
    this.onSession = opts.onSession ?? null;
  }

  setTurnContext(channelId: string | null, threadRootId: string | null): void {
    this.channelId = channelId;
    this.threadRootId = threadRootId;
  }

  resetPhase(): void {
    this.phase = null;
  }

  /** 管理器喂入一个 turn 前调用：标记 turn 进行中（init→idle 抑制窗口）。 */
  beginTurn(): void {
    this.turnActive = true;
  }

  /** 一次 spawn 的运行态复位（confirmed/turn 标记；sessionId 保留）。 */
  resetRun(): void {
    this.confirmed = false;
    this.turnActive = false;
    this.phase = null;
  }

  // ------------------------------------------------------------ 分发

  /** 路由一帧（防腐：任何未知/畸形帧不得抛出到调用方）。 */
  async process(frame: JsonObject): Promise<void> {
    const ftype = frame['type'];
    if (ftype === 'system') {
      await this.onSystem(frame);
    } else if (ftype === 'stream_event') {
      await this.onStreamEvent(frame);
    } else if (ftype === 'assistant') {
      await this.onAssistant(frame);
    } else if (ftype === 'user') {
      await this.onUser(frame);
    } else if (ftype === 'result') {
      await this.onResult(frame);
    } else {
      // py str(None)="None" ↔ TS String(undefined)="undefined"：未知键的字面拼写随宿主语言（登记差异）
      this.countUnknown(String(ftype));
    }
  }

  // ------------------------------------------------------------ system（init → idle）

  private async onSystem(frame: JsonObject): Promise<void> {
    const subtype = frame['subtype'];
    if (subtype === 'init') {
      const sid = frame['session_id'] as string | null | undefined;
      if (sid && sid !== this.sessionId) {
        this.sessionId = sid;
        if (this.onSession !== null) {
          this.onSession(sid); // 会话簿记（daemon/state/<id>.json，§4）
        }
      }
      this.model = (frame['model'] as string | null | undefined) || this.model;
      this.confirmed = true; // resume 生效 / 会话就绪的权威凭据
      if (!this.turnActive) {
        // turn 未在跑（罕见：某些模式 init 先于输入）→ 就绪 idle。
        await this.status('idle');
      }
    } else {
      // system/status | api_retry | notification（契约外噪声）→ 防腐层
      this.countUnknown(`system/${String(subtype)}`);
    }
  }

  // ------------------------------------------------------------ stream_event（相位聚合）

  private async onStreamEvent(frame: JsonObject): Promise<void> {
    const event = (frame['event'] || {}) as JsonObject;
    const etype = event['type'];
    if (etype === 'content_block_start') {
      const block = (event['content_block'] || {}) as JsonObject;
      const btype = block['type'];
      if (btype === 'thinking') {
        await this.switchPhase(P_THINKING);
      } else if (btype === 'text') {
        await this.switchPhase(P_REPLYING);
      } else if (btype === 'tool_use') {
        const name = (block['name'] as string | undefined) || '';
        const bid = block['id'] as string | undefined;
        if (bid) {
          this.toolUses.set(bid, [name, (block['input'] || {}) as JsonObject]);
        }
        await this.switchPhase(phaseForTool(name));
      }
    }
    // message_start / content_block_delta / content_block_stop / message_delta /
    // message_stop：同相位或无相位 → 不上报（§7.2）
  }

  private async switchPhase(phase: string): Promise<void> {
    if (phase === this.phase) {
      return;
    }
    this.phase = phase;
    await this.sink.onActivity(this.agentMemberId, phase);
  }

  // ------------------------------------------------------------ assistant（turn_output 诊断）

  private async onAssistant(frame: JsonObject): Promise<void> {
    const msg = (frame['message'] || {}) as JsonObject;
    const content = (msg['content'] || []) as JsonObject[];
    const textParts: string[] = [];
    let toolCalls = 0;
    for (const block of content) {
      const btype = block['type'];
      if (btype === 'text') {
        textParts.push((block['text'] as string | undefined) || '');
      } else if (btype === 'tool_use') {
        toolCalls += 1;
        const bid = block['id'] as string | undefined;
        if (bid) {
          this.toolUses.set(bid, [
            (block['name'] as string | undefined) || '',
            (block['input'] || {}) as JsonObject,
          ]);
        }
      }
    }
    const preview = textParts.join('').slice(0, TURN_OUTPUT_PREVIEW_MAX);
    this.diag('agent.turn_output', {
      preview, // 正文**不外发**，仅截断留痕（铁律 5）
      tool_calls: toolCalls,
      stop_reason: msg['stop_reason'] ?? null,
    });
  }

  // ------------------------------------------------------------ user（tool_result → 诊断）

  private async onUser(frame: JsonObject): Promise<void> {
    const msg = (frame['message'] || {}) as JsonObject;
    const content = msg['content'] || [];
    if (!Array.isArray(content)) {
      return;
    }
    for (const block of content as unknown[]) {
      if (!isDict(block) || block['type'] !== 'tool_result') {
        continue;
      }
      const tid = String(block['tool_use_id'] || ''); // 缺失/非法 → "" 查默认（帧防腐）
      const [name, toolInput] = this.toolUses.get(tid) ?? ['', {} as JsonObject];
      const isError = Boolean(block['is_error']);
      if (TOOLS_COMMAND.has(name)) {
        const cmd = String(toolInput['command'] ?? '').slice(0, COMMAND_PREVIEW_MAX);
        this.diag('agent.command', { command: cmd, is_error: isError });
      } else if (TOOLS_WRITING.has(name)) {
        const kind = name === 'Write' ? 'create' : 'edit';
        this.diag('agent.file_edit', {
          path: toolInput['file_path'] ?? null,
          kind,
          is_error: isError,
        });
      } else {
        this.diag('agent.tool_call', { tool: name || 'unknown', ok: !isError });
      }
    }
  }

  // ------------------------------------------------------------ result（usage + idle/error）

  private async onResult(frame: JsonObject): Promise<void> {
    const usage = (frame['usage'] || {}) as JsonObject;
    const event: TokenUsageEventIn = {
      id: this.ulid(), // 每 result 帧一个 ULID（exactly-once 去重根基，§7.4）
      agent_member_id: this.agentMemberId,
      channel_id: this.channelId,
      thread_root_id: this.threadRootId,
      input_tokens: asInt(usage['input_tokens']),
      output_tokens: asInt(usage['output_tokens']),
      cache_read_tokens: asInt(usage['cache_read_input_tokens']),
      cache_write_tokens: asInt(usage['cache_creation_input_tokens']),
      source_session: this.sessionId,
      reported_at: this.now(),
    };
    this.sink.onUsage(event); // total_cost_usd / modelUsage.costUSD 忽略（永不折算货币）

    this.confirmed = true;
    this.turnActive = false;
    this.resetPhase();
    this.toolUses.clear();
    const subtype = frame['subtype'];
    const isError = Boolean(frame['is_error']) || Boolean(frame['api_error_status']);
    if (subtype === 'success' && !isError) {
      await this.status('idle');
    } else {
      const detail = String(subtype || frame['api_error_status'] || 'result_error');
      await this.status('error', detail);
    }
  }

  // ------------------------------------------------------------ 防腐层

  private countUnknown(key: string): void {
    const seen = this.unknownCounts[key] ?? 0;
    this.unknownCounts[key] = seen + 1;
    if (seen === 0) {
      // 每种未知类型首现写一条低频诊断，后续静默累加
      this.diag('agent.unknown_frame', { type: key, count: 1 });
    }
  }

  // ------------------------------------------------------------ 回调底座

  private async status(status: AgentStatus, errorDetail: string | null = null): Promise<void> {
    this.lastStatus = status;
    await this.sink.onStatusChanged(this.agentMemberId, status, errorDetail);
  }

  private diag(dtype: string, payload: JsonObject): void {
    this.sink.onDiagnostic({
      agent_member_id: this.agentMemberId,
      type: dtype,
      channel_id: this.channelId,
      // py 侧 pydantic 默认字段（task_id/batch_id=None）显式兑现——model_dump(mode="json")
      // 序列化含 null 键，TS 对象与之逐键对齐（W1 buffer JSONL 形状对等）。
      task_id: null,
      batch_id: null,
      payload,
      at: this.now(),
    });
  }
}
